staging/src/k8s.io/kubectl/pkg/kuberc/kuberc.go

```diff
@@ -0,0 +1,315 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kuberc
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/homedir"
+	"k8s.io/kubectl/pkg/cmd/util"
+	"k8s.io/kubectl/pkg/config/v1alpha1"
+	"k8s.io/kubectl/pkg/scheme"
+)
+
+const RecommendedKubeRCFileName = ".kuberc"
+
+var (
+	RecommendedConfigDir  = filepath.Join(homedir.HomeDir(), clientcmd.RecommendedHomeDir)
+	RecommendedKubeRCFile = filepath.Join(RecommendedConfigDir, RecommendedKubeRCFileName)
+)
+
+// PreferencesHandler is responsible for setting default flags
+// arguments based on user's kuberc configuration.
+type PreferencesHandler interface {
+	AddFlags(flags *pflag.FlagSet)
+	ApplyOverrides(rootCmd *cobra.Command, args []string, errOut io.Writer) error
+	ApplyAliases(rootCmd *cobra.Command, args []string, errOut io.Writer) ([]string, error)
+}
+
+// Preferences stores the kuberc file coming either from environment variable
+// or file from set in flag or the default kuberc path.
+type Preferences struct {
+	GetPreferencesFunc func(kuberc string) (*v1alpha1.Preferences, error)
+
+	aliases map[string]struct{}
+}
+
+// NewPreferences returns initialized Prefrences object.
+func NewPreferences() PreferencesHandler {
+	return &Preferences{
+		GetPreferencesFunc: DefaultGetPreferences,
+		aliases:            make(map[string]struct{}),
+	}
+}
+
+// AddFlags adds kuberc related flags into the command.
+func (p *Preferences) AddFlags(flags *pflag.FlagSet) {
+	if util.KubeRC.IsEnabled() {
+		flags.String("kuberc", "", "Path to the kuberc file to use for preferences.")
+	}
+}
+
+// ApplyOverrides finds the command and sets the defaulted flag values in kuberc.
+func (p *Preferences) ApplyOverrides(rootCmd *cobra.Command, args []string, errOut io.Writer) error {
+	if len(args) <= 1 {
+		return nil
+	}
+
+	if !util.KubeRC.IsEnabled() {
+		return nil
+	}
+
+	kubercPath := getExplicitKuberc(args)
+	kuberc, err := p.GetPreferencesFunc(kubercPath)
+	if err != nil {
+		return fmt.Errorf("kuberc error %w", err)
+	}
+
+	if kuberc == nil {
+		return nil
+	}
+
+	args = args[1:]
+	cmd, _, err := rootCmd.Find(args)
+	if err != nil {
+		fmt.Fprintf(errOut, "command not found %v\n", err)
+		return nil
+	}
+
+	for _, c := range kuberc.Spec.Overrides {
+		parsedCmds := strings.Split(c.Command, " ")
+		overrideCmd, _, err := rootCmd.Find(parsedCmds)
+		if err != nil {
+			// this may be referring to the alias command which is not initialized
+			// because the actual command is totally different.
+			return nil
+		}
+		if overrideCmd.Name() != cmd.Name() {
+			continue
+		}
+
+		if _, ok := p.aliases[cmd.Name()]; ok {
+			return fmt.Errorf("alias %s can not be overridden", cmd.Name())
```

Comment:
I was going to comment "Similar to the above comment. Print a warning and continue"

But I think it could be dangerous to continue executing a command you expected to be overridden.

So I think the code here is fine how it is now.

Reply:
Yes, due to security concerns, we can't allow it or iterate over by simple warning message.